kotlinx-coroutines-core/jvm/src/scheduling/WorkQueue.kt

```diff
@@ -138,23 +153,41 @@ internal class WorkQueue {
             } // Failed -> someone else stole it
         }
 
+        return pollWithMode(onlyBlocking = true /* only blocking */)
+    }
+
+    fun pollCpu(): Task? {
+        while (true) { // Poll the slot
+            val lastScheduled = lastScheduledTask.value ?: break
+            if (lastScheduled.isBlocking) break
+            if (lastScheduledTask.compareAndSet(lastScheduled, null)) {
+                return lastScheduled
+            } // Failed -> someone else stole it
+        }
+
+        return pollWithMode(onlyBlocking = false /* only cpu */)
+    }
+
+    private fun pollWithMode(/* Only blocking OR only CPU */onlyBlocking: Boolean): Task? {
         val start = consumerIndex.value
         var end = producerIndex.value
-
-        while (start != end && blockingTasksInBuffer.value > 0) {
-            val task = tryExtractBlockingTask(--end)
+        // CPU or (BLOCKING & hasBlocking)
+        val shouldProceed = !onlyBlocking || blockingTasksInBuffer.value > 0
+        while (start != end && shouldProceed) {
```

Comment:
`shouldProceed` doesn't update. It doesn't feel like it's intentional, but if it is, it can be moved to an `if` wrapping the `while`.